url/url.go

```diff
@@ -0,0 +1,189 @@
+/* package URL handles Helm-DM URLs
+
+Helm uses three kinds of URLs:
+
+	- Fully qualified (long) names: https://example.com/foo/bar-1.2.3.tgz
+	- Short names: helm:example.com/foo/bar#1.2.3
+	- Local names: file:///foo/bar
+
+This package provides utilities for working with this type of URL.
+*/
+package url
+
+import (
+	"errors"
+	"fmt"
+	"net/url"
+	"path/filepath"
+	"regexp"
+	"strings"
+)
+
+const (
+	SchemeHTTP  = "http"
+	SchemeHTTPS = "https"
+	SchemeHelm  = "helm"
+	SchemeFile  = "file"
+)
+
+// TarNameRegex parses the name component of a URI and breaks it into a name and version.
+//
+// This borrows liberally from github.com/Masterminds/semver.
+const TarNameRegex = `([0-9A-Za-z\-_/]+)-(v?([0-9]+)(\.[0-9]+)?(\.[0-9]+)?` +
+	`(-([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?` +
+	`(\+([0-9A-Za-z\-]+(\.[0-9A-Za-z\-]+)*))?)(.tgz)?`
+
+var tnregexp *regexp.Regexp
+
+func init() {
+	tnregexp = regexp.MustCompile("^" + TarNameRegex + "$")
+}
+
+type URL struct {
```

Comment:
Because we don't want those features, right? Example: Bucket is not a part of a URL. And depending on wheter it's a short or long URL, Bucket is stored in a different part of the URL (the opaque vs the path). Probably, we should just not be calling this a URL, since semantically it's not. It's a _something_ that can be represented using different URL forms. Any idea for a better name?